components/server/src/orgs/organization-service.ts

```diff
@@ -0,0 +1,83 @@
+import { ResponseError } from "vscode-ws-jsonrpc";
+import { ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
+import { OrganizationPermission } from "../authorization/definitions";
+
+@injectable()
+export class OrganizationService {
+    constructor(
+        @inject(TeamDB) private readonly teamDB: TeamDB,
+        @inject(Authorizer) private readonly auth: Authorizer,
+    ) {}
+    /**
+     * createOrganization creates a new organization
+     */
+    async createOrganization(userId: string, name: string): Promise<Organization> {
+        let result: Organization;
+        try {
+            result = await this.teamDB.transaction(async (db) => {
+                result = await db.createTeam(userId, name);
+                await this.auth.addOrganizationOwnerRole(result.id, userId);
+                return result;
+            });
+        } catch (err) {
+            if (result! && result.id) {
+                await this.auth.removeUserFromOrg(result.id, userId);
+            }
+
+            throw err;
+        }
+        return result;
+    }
+
+    /**
+     * getGenericInvite returns the generic invite for the given organization or creates one if no invite exists.
+     *
+     * @param userId
+     * @param orgID
+     * @returns
+     */
+    public async getGenericInvite(userId: string, orgID: string): Promise<TeamMembershipInvite> {
+        const invite = await this.teamDB.findGenericInviteByTeamId(orgID);
+        if (invite) {
+            if (await this.teamDB.hasActiveSSO(orgID)) {
+                throw new ResponseError(ErrorCodes.NOT_FOUND, "Invites are disabled for SSO-enabled organizations.");
+            }
+            await this.checkPermissionAndThrow(userId, "write_members", orgID);
+            return invite;
+        }
+        return this.resetGenericInvite(userId, orgID);
+    }
+
+    public async resetGenericInvite(userId: string, orgID: string): Promise<TeamMembershipInvite> {
+        if (await this.teamDB.hasActiveSSO(orgID)) {
+            throw new ResponseError(ErrorCodes.NOT_FOUND, "Invites are disabled for SSO-enabled organizations.");
+        }
+        await this.checkPermissionAndThrow(userId, "write_members", orgID);
+        return this.teamDB.resetGenericInvite(orgID);
+    }
+
+    private async checkPermissionAndThrow(userId: string, permission: OrganizationPermission, orgID: string) {
+        if (await this.auth.hasPermissionOnOrganization(userId, permission, orgID)) {
+            return;
+        }
+        // check if the user has read permission
+        if ("read_info" === permission || !(await this.auth.hasPermissionOnOrganization(userId, "read_info", orgID))) {
+            throw new ResponseError(ErrorCodes.NOT_FOUND, `Organization ID: ${orgID} not found.`);
+        }
+
+        throw new ResponseError(
+            ErrorCodes.PERMISSION_DENIED,
+            "You do not have the required permission to perform this operation.",
```

Comment:
Could we be more specific in here? Something like `You do not have ${permission} on ${orgID}`?